ENIGMAsystem/SHELL/Platforms/SDL/Joystick.cpp

```diff
@@ -0,0 +1,106 @@
+/** Copyright (C) 2018 Samuel Venable
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "SDL2/SDL.h"
+#include "Joystick.h"
+#include "Platforms/General/PFjoystick.h"
+#include <string>
+
+namespace {
+
+SDL_Joystick *joystick1;
+SDL_Joystick *joystick2;
+bool update = true;
+
+static inline SDL_Joystick* joystick_get_handle(int id) {
+  return (id - 1 < 0.5) ? joystick1 : joystick2;
+}
+
+} // anonymous namespace
+
+namespace enigma_user {
+
+bool joystick_exists(int id) {
+  return (SDL_JoystickGetAttached(joystick_get_handle(id)) == SDL_TRUE);
+}
+
+std::string joystick_name(int id) {
+  const char *name;
+  name = SDL_JoystickName(joystick_get_handle(id));
+  return (name ? name : "");
+}
+
+int joystick_axes(int id) {
+  int axes;
+  SDL_Joystick *joystick = joystick_get_handle(id);
+  axes = SDL_JoystickNumAxes(joystick);
+  return (axes > 0 ? axes : 0);
+}
+
+int joystick_buttons(int id) {
+  int buttons;
+  SDL_Joystick *joystick = joystick_get_handle(id);
+  buttons = SDL_JoystickNumButtons(joystick);
+  return (buttons > 0 ? buttons : 0);
+}
+
+bool joystick_button(int id, int numb) {
+  return SDL_JoystickGetButton(joystick_get_handle(id), numb - 1);
```

Comment:
This is just a note, we may need a map here later, I see in testing that `joystick_button(1, 0)` returns `true` when a controller is plugged in and no buttons are pressed.

Reply:
@RobertBColton we could just prevent the user from entering something less than 1, by checking if less than 1, and then if so, forcing it to equal 1.

Reply:
Are you sure it's all just shifted by 1 button though? I can test and compare it to GM8. Because what does SDL think button 0 is? If SDL's button 0 maps to GML's button 32 then you would need `(numb - 1) % 32`.

Reply:
Just leave it for now, address it later if and when somebody complains.